add --trim option for generating smaller binaries

JeffBezanson opened the pull request:
This adds a command line option `--trim` that builds images where code is only included if it is statically reachable from methods marked using the new function `entrypoint`. Compile-time errors are given for call sites that are too dynamic to allow trimming the call graph (however there is an `unsafe` option if you want to try building anyway to see what happens).

The PR has two other components. One is changes to Base that generally allow more code to be compiled in this mode. These changes will either be merged in separate PRs or moved to a separate part of the workflow (where we will build a custom system image for this purpose). The branch is set up this way to make it easy to check out and try the functionality.

The other component is everything in the `juliac/` directory, which implements a compiler driver script based on this new option, along with some examples and tests. This will eventually become a package "app" that depends on PackageCompiler and provides a CLI for all of this stuff, so it will not be merged here. To try an example:

```
julia contrib/juliac.jl --output-exe hello --trim test/trimming/hello.jl
```

When stripped the resulting executable is currently about 900kb on my machine.

Also includes a lot of work by @topolarity

gbaraldi committed:
Implement changes to the runtime that allow for building small binaries,
also implement a --strict mode where generating a dynamic dispatch becomes a compile time error

gbaraldi committed:
Implement example driver and simple tests

gbaraldi committed:
Less invasive changes to base to make hello world work

gbaraldi committed:
Fix some typos and refactor cli usage to only use static-call-graph. Also refactor the runtime to only have one option

JeffBezanson committed:
make argument to --static-call-graph optional, improve help text

JeffBezanson committed:
misc formatting and cleanup

gbaraldi committed:
Check all builtins for potential dynamic dispatches

gbaraldi committed:
Fix typos in the driver and exit with exit(1)

gbaraldi committed:
Make stacktrace printing robust and add overload to init_project to avoid logging

JeffBezanson committed:
fix warning for print_retainers by ifdef'ing it out

gbaraldi committed:
Save current state of the branch.
Union split test gives the wrong result

JeffBezanson committed:
WIP

update rebuild_methtables to hopefully work with abstract signatures

gbaraldi committed:
Add default --static-call-graph option

JeffBezanson committed:
don't compile for typeinf world with static-call-graph

JeffBezanson committed:
rename option to --trim

ViralBShah commented:
```
➜  juliac git:(jb/gb/static-call-graph) ✗ ../julia juliac.jl --output-exe hello --trim exe_examples/hello_world.jl
ld: warning: ignoring duplicate libraries: '-ljulia'
ld: warning: reexported library with install name '@rpath/libunwind.1.dylib' found at '/Users/viral/julia/usr/lib/libunwind.1.0.dylib' couldn't be matched with any parent library and will be linked directly
run(`cc $(allflags) -o $outname -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path $(julia_libs)`) = Process(`cc -std=gnu11 -I/Users/viral/julia/usr/include/julia -fPIC -L/Users/viral/julia/usr/lib -L/Users/viral/julia/usr/lib/julia -Wl,-rpath,/Users/viral/julia/usr/lib -ljulia -o hello -Wl,-all_load /var/folders/qj/54cx7lcs2qv9z1b4f_110xb00000gn/T/jl_fPbFYU/img.a -Wl, /var/folders/qj/54cx7lcs2qv9z1b4f_110xb00000gn/T/jl_fPbFYU/init.a -ljulia -ljulia-internal`, ProcessExited(0))
➜  juliac git:(jb/gb/static-call-graph) ✗ ls -l hello
-rwxr-xr-x  1 viral  staff  1082296 Jul  6 09:57 hello
```

JeffBezanson committed:
wip

gbaraldi committed:
Fix unsafe-warn mutiple stacktrace output

JeffBezanson committed:
fixes for threads and tasks with trimming

timholy commented:
#55104 gets the lib example working

MasonProtter commented:
Let me know if issues should be raised elsewhere, but I ran into a problem where the compiled binaries don't seem to get `ARGS`. Here's a MWE: (`args.jl`)
```julia
module ArgsDemo

Base.@ccallable function main()::Cint
    ccall(:jl_, Cvoid, (Any,), ARGS)
    return 0
end

end
```
```
❯ ~/julia/./julia --startup=no ~/julia/juliac/juliac.jl --output-exe args --trim args.jl
run(`cc $(allflags) -o $outname -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path $(julia_libs)`) = Process(`cc -std=gnu11 -I/home/masonp/julia/usr/include/julia -fPIC -L/home/masonp/julia/usr/lib -Wl,--export-dynamic -L/home/masonp/julia/usr/lib/julia -Wl,-rpath,/home/masonp/julia/usr/lib -Wl,-rpath,/home/masonp/julia/usr/lib/julia -ljulia -o args -Wl,--whole-archive /tmp/jl_2vsSFl/img.a -Wl,--no-whole-archive /tmp/jl_2vsSFl/init.a -ljulia -ljulia-internal`, ProcessExited(0))

❯ ./args
Array{String, 1}(dims=(0,), mem=Memory{String}(8, 0x5589c120d310)[#<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>])

❯ ./args arg1 arg2 agr3
Array{String, 1}(dims=(0,), mem=Memory{String}(8, 0x558aac9f6310)[#<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>, #<null>])
```

timholy committed:
Get lib example working, create build script (#55104)

Note this is a PR against #55047

Co-authored by: Gabriel Baraldi <baraldigabriel@gmail.com>

JeffBezanson commented:
That should be fixed by integration with PackageCompiler. It knows how to do all that stuff.

JeffBezanson committed:
fix whitespace

JeffBezanson committed:
convert build scripts to a test case for now

JeffBezanson commented:
OK, I deleted everything. Don't worry it's on a backup branch jb/gb/static-call-graph-backup. I converted the juliac machinery into a test case so we can focus here on just merging the core compiler functionality.

Now there are only a tiny number of controversial/weird changes:
- unoptimize_throw_blocks: Can possibly be changed upstream? I believe we want to do that anyway. Otherwise the test case might need to build its own sysimage.
- Linalg init changes: these are weird error cases and I think can just be changed not to use logging?
- Setting max_args of printing functions: weird but harmless; can be upstreamed?

JeffBezanson committed:
fix warning printing in LinearAlgebra

JeffBezanson committed:
fix _reinit_stdio needed for codegen error printing

JeffBezanson committed:
rename --static-call-graph to --trim

add NEWS and docs

JeffBezanson committed:
fix analyzegc

KristofferC commented on NEWS.md:
```diff
@@ -4,6 +4,8 @@ Julia v1.12 Release Notes
 New language features
 ---------------------
 
+- New option `--trim` for building "trimmed" binaries, where code not provably reachable from entry points
```
```suggestion
- New option `--trim` for building "trimmed" binaries, where code provably not reachable from entry points
```

To my non-native brain, this sounds better

KristofferC commented:
Does it make sense to add `depwarn(msg, funcsym; force::Bool=false) = nothing` to `buildscript.jl`? Right now a `depwarn` calls will error.

It could be possible to avoid this by constricting the input types of `Base._depwarn` in combination with https://github.com/JuliaLang/julia/pull/55166 but it feels unlikely you want to see any depwarns after statically compiling something.

MasonProtter commented:
Since we have a static call graph here, maybe we should just hoist the depwarns to be evaluated as we traverse the call graph?

In most ahead-of-time compiled languages, there's large classes of warnings and errors that are emitted during compilation, rather than from the compiled binary itself.

mbauman commented:
depwarns are often statically resolvable (and currently all are), but in the past they have sometimes required runtime information to know if they should even exist in the first place. E.g., 

https://github.com/JuliaLang/julia/blob/v0.7.0/base/deprecated.jl#L525-L528

MasonProtter commented:
Not that it's actually relevant to your wider point, but I think that depwarn *is* statically resolvable because it's checking if a `Pairs{<:NamedTuple}` is empty.

But for cases where someone really does have a dynamic depwarn, I honestly think it's fine to emit it anyways, even if it's only hit on one branch. That's the same for how we'd error out if someone did
```
function f(x)
    if rand() == 100
        something_type_stable(x)
    else
        something_type_unstable(x)
    end
```

JeffBezanson committed:
fixes

JeffBezanson commented:
Hopefully there aren't many examples out there of dynamic depwarns. But it could be a problem because people won't want false positives from those. Anyway `depwarn` is currently not known to the compiler, but I guess we can hack it in if we want.

KristofferC commented:
W.r.t depwarn, the reason I mentioned it was just because I first hit and error here

https://github.com/JuliaLang/julia/blob/86231ce5763a41a6661d7834a28ad1c37526044a/base/deprecated.jl#L257

due to an `invokelatest` and untyped arguments and when I made the arguments concrete I hit the dynamic logging later in the function. Special casing it for now just felt easier.

JeffBezanson commented:
Yes I agree the right thing for now is just to disable it.

JeffBezanson committed:
fix for warning test

timholy commented on src/staticdata.c:
```diff
@@ -751,30 +757,98 @@ static uintptr_t jl_fptr_id(void *fptr)
+                    // ... point to modules ...
+                    (jl_is_module(jl_atomic_load_relaxed(&b->value)) ||
+                     // ... or point to __init__ methods ...
+                     !strcmp(jl_symbol_name(b->globalref->name), "__init__") ||
```
You have my condolences 😆

timholy commented on src/staticdata.c:
```diff
@@ -2726,14 +2898,14 @@ static void jl_save_system_image_to_stream(ios_t *f, jl_array_t *mod_array,
             for (i = 0; tags[i] != NULL; i++) {
                 jl_value_t *tag = *tags[i];
-                jl_queue_for_serialization(&s, tag);
+                jl_queue_for_serialization(&s, tag, /* retainer */ NULL);
```
So the idea is that you enable debugging from individual sites? I was expecting that one would always pass `retainer` but that it might be NULL if you're not debugging. It might be important to explain this wherever `DEBUG_RETAINERS` gets documented.

timholy commented on test/trimming/buildscript.jl:
```diff
@@ -0,0 +1,277 @@
+output_type = ARGS[2]
+add_ccallables = ARGS[3] == "true"
+
+# Initialize some things not usually initialized when output is request
```
```suggestion
# Initialize some things not usually initialized when output is requested
```

timholy commented on test/trimming/buildscript.jl:
```diff
@@ -0,0 +1,277 @@
+            print(io, "?")
+        end
+    end
+    function show(io::IO, T::Type)
```
```suggestion
    function show(io::IO, @nospecialize(T::Type))
```
Presumably this happens already due to heuristics, but I wonder if we'll want to experiment with changing those heuristics for static builds---there are some things you might want to specialize that we don't ordinarily specialize.

JeffBezanson replied:
Yes we will want to do that.

timholy reviewed (commented):
Very exciting, really looking forward to this!

oscardssmith commented on doc/src/devdocs/sysimg.md:
```diff
@@ -117,3 +117,68 @@ See code comments for each components for more implementation details.
+image built, but it may crash at run time if needed code is not present.
+
+It typically makes sense to specify `--strip-ir` and `--strip-metadata` along with `--trim`,
+since those options remove information not needed in a small binary meant for deployment.
```
yeah the main reason not to is it will make debugging a lot harder.

timholy replied:
Worth spelling this out then.

JeffBezanson committed:
Update doc/src/devdocs/sysimg.md

Co-authored-by: Tim Holy <tim.holy@gmail.com>

JeffBezanson committed:
Update doc/src/devdocs/sysimg.md

Co-authored-by: Tim Holy <tim.holy@gmail.com>

JeffBezanson committed:
Update doc/src/manual/command-line-interface.md

Co-authored-by: Tim Holy <tim.holy@gmail.com>

JeffBezanson committed:
Update test/trimming/buildscript.jl

Co-authored-by: Tim Holy <tim.holy@gmail.com>

JeffBezanson commented on src/staticdata.c:
```diff
@@ -751,30 +757,98 @@ static uintptr_t jl_fptr_id(void *fptr)
+static void jl_queue_for_serialization_(jl_serializer_state *s, jl_value_t *v, jl_value_t *p, int recursive, int immediate) JL_GC_DISABLED;
+
+#ifdef DEBUG_RETAINERS
+static void print_retainers(jl_value_t *v) JL_GC_DISABLED
```
I think this can just be deleted.

JeffBezanson committed:
trimming doc clarifications

JeffBezanson committed:
rm print_retainers debug code

JeffBezanson committed:
hook up CI for trimming

JeffBezanson commented:
First CI of a trimmed binary! :tada: :raised_hands:

JeffBezanson committed:
allow single non-concrete method match in static call graph

topolarity committed:
Add missing `jl_emit_codeinst` failure check

JeffBezanson committed:
add juliac.jl to contrib and install it
hook up tests better

JeffBezanson commented:
Now with `contrib/juliac.jl`! I wasn't sure where to put this script but `contrib` seems like a good place for now. The script is experimental but this branch is now pretty complete and usable.

vchuravy commented:
> I wasn't sure where to put this script but contrib seems like a good place for now. The script is experimental but this branch is now pretty complete and usable.

Stdlib whence https://github.com/JuliaLang/Pkg.jl/pull/3772 lands?

vchuravy commented on base/experimental.jl:
```diff
@@ -457,4 +457,18 @@ without adding them to the global method table.
 :@MethodTable
 
+"""
+    Base.Experimental.entrypoint(f, argtypes::Tuple)
```
What is the difference between `entrypoint` and `precompile`?
Could we just use `precompile`, or add a keyword to it?

The docstring here should include the "where code not provably reachable from entry points is removed." bit

JeffBezanson replied:
The problem is we want to mark only the minimal needed entry points, and not compile anything speculatively, so existing precompile statements should be ignored. I guess it could be a keyword argument, but only in the sense that any function can be turned into any other function by adding a keyword argument :shrug:

vchuravy commented on src/aotcompile.cpp:
```diff
@@ -284,6 +295,7 @@ jl_code_instance_t *jl_ci_cache_lookup(const jl_cgparams_t &cgparams, jl_method_
     jl_value_t *ci = cgparams.lookup(mi, world, world);
     JL_GC_PROMISE_ROOTED(ci);
     jl_code_instance_t *codeinst = NULL;
+    JL_GC_PUSH1(&codeinst);
```
This looks like a proper bugfix? Shouldn't the verifier have noticed this?

JeffBezanson replied:
Yes --- on this branch analyzegc complains about this, but on master it doesn't and I don't know why.

maleadt replied:
IIRC the analyzer is heuristically driven, so the changes here must have changed how it visits the codebase.

Keno replied:
I don't see why this root is required.

vchuravy commented on src/aotcompile.cpp:
```diff
@@ -373,30 +391,63 @@ void *jl_create_native_impl(jl_array_t *methods, LLVMOrcThreadSafeModuleRef llvm
                 continue;
             }
             mi = (jl_method_instance_t*)item;
+            compile_mi:
```
```suggestion
compile_mi:
```

Or indent the rest? The label was difficult to spot the first go-around, and I agree we should refactor the code to avoid the goto.

gbaraldi commented on src/codegen.cpp:
```diff
@@ -2122,6 +2115,179 @@ static jl_cgval_t emit_invoke(jl_codectx_t &ctx, const jl_cgval_t &lival, ArrayR
+
+static jl_array_t* build_stack_crumbs(jl_codectx_t &ctx) JL_NOTSAFEPOINT
+{
+    static intptr_t counter = 5;
```
It just serves a random sentinel that is never going to be a valid pointer. It could be any number but that can't be misconstrued as a pointer. And it has to be unique in the map per `build_stack_crumbs` call. Maybe there's a nicer implementation of this but I struggled to come up with one.

Keno commented on .buildkite-external-version:
```diff
@@ -1 +1 @@
-main
+jb/trim_ci_job
```
Should be on main for the PR merge and then a separate merge to main in the PR repo

Keno commented on contrib/juliac.jl:
```diff
@@ -0,0 +1,110 @@
+julia_libs = Base.shell_split(Base.isdebugbuild() ? "-ljulia-debug -ljulia-internal-debug" : "-ljulia -ljulia-internal")
+try
+    if output_type == "--output-lib"
+        run(`cc $(allflags) -o $outname -shared -Wl,$(Base.Linking.WHOLE_ARCHIVE) $img_path  -Wl,$(Base.Linking.NO_WHOLE_ARCHIVE) $init_path  $(julia_libs)`)
```
Should use the linker we ship rather than trying system linker

JeffBezanson replied:
Where is it?

gbaraldi replied:
It's in Base.Linking. Though for now we continue to need a C compiler. I think if we want to get fancy we can ship some LLVM bitcode and ask the LLVM compiler we ship with julia to compile it to a object file

Keno commented on contrib/juliac.jl:
```diff
@@ -0,0 +1,110 @@
+open(initsrc_path, "w") do io
+    print(io, """
+              #include <julia.h>
+              __attribute__((constructor)) void static_init(void) {
```
Why can this not be written in Julia?

JeffBezanson replied:
I think because julia code cannot run before `julia_init` is called.

Keno commented:
Feels very jammed in in places, but I think we can get this in and then clean up later.

vtjnash commented on contrib/juliac.jl:
```diff
@@ -0,0 +1,110 @@
+              __attribute__((constructor)) void static_init(void) {
+                  if (jl_is_initialized())
+                      return;
+                  julia_init(JL_IMAGE_IN_MEMORY);
```
I am pretty sure this does many things that are forbidden in a `__attribute__((constructor))` function (e.g. accessing C++ is potentially not legal, as their constructors might not have run yet?), so we need to split this function between things that are legal and must be done, and things that can be done a bit later

gbaraldi replied:
Me and @topolarity have talked about moving this to the adopt_thread call. And put this around a pthread_once

giordano commented:
Unfortunately this build system doesn't seem to be reproducible.  When recompiling the same source code from scratch (delete the intermediate object files between reruns, e.g. `make -BC test/trimming JULIA=$(realpath usr/bin/julia) BIN=$(realpath usr/bin)` for the built-in example) produces two different binaries, even with different sizes, which makes validation of the output hard.  Here's a [diffoscope](https://try.diffoscope.org/exrapftkepxg.html) of two builds of the same hello world example

topolarity commented:
> Unfortunately this build system doesn't seem to be reproducible. 

Have our sysimage builds ever been reproducible? This functionality uses the existing sysimage generation directly

JeffBezanson commented:
Yes, I think sysimage builds have always been like this and it should be a separate item to work on.

JeffBezanson commented:
I think this is mergeable now so we can get the --trim option in. The only difficulties are basically in the tests and docs, which can be updated any time.

PallHaraldsson commented:
Please merge this for 1.12. I'm very exited to try this (more) easily out, and for regular users to try, so would it be possible to add this to 1.11.0 or at least some 1.11.x point release (marked as an experimental new feature)? It seems this is largely not affecting regular Julia execution, lives in contrib, but I do see though some changes in codegen and staticdata.c, so I'm not sure how invasive this is, and I understand if this is to late considering RC4...

MasonProtter commented:
The v1.11 feature freeze was over half a year ago.
